### Week 2 task by Denis Khryashchev

In [1]:
# Importing the Apriori implementation of mlxtend library
from mlxtend.frequent_patterns import apriori
from mlxtend.preprocessing import OnehotTransactions
import numpy as np
import pandas as pd
import csv

# Setting pandas to display wider columns (up to 150 characters max per column)
pd.set_option("display.max_colwidth", 250)

In [2]:
# Loading the dataset
# The data were accessed from https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/
#filename = "agaricus-lepiota.data"
filename = "expanded.csv"
with open(filename, 'rb') as csvfile:
    reader = csv.reader(csvfile, delimiter=',')
    mushrooms = map(lambda mushroom: mushroom, reader)

In [3]:
# Field names from https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.names
colnames = ['classes', 'cap-shape', 'cap-surface', 'cap-color', 'bruises?', 'odor', 'gill-attachment', 'gill-spacing',
            'gill-size', 'gill-color', 'stalk-shape', 'stalk-root', 'stalk-surface-above-ring', 'stalk-surface-below-ring', 
            'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number', 'ring-type',
            'spore-print-color', 'population', 'habitat']

In [4]:
# Pairing example
zip(colnames, mushrooms[0])

[('classes', 'EDIBLE'),
 ('cap-shape', 'CONVEX'),
 ('cap-surface', 'SMOOTH'),
 ('cap-color', 'WHITE'),
 ('bruises?', 'BRUISES'),
 ('odor', 'ALMOND'),
 ('gill-attachment', 'FREE'),
 ('gill-spacing', 'CROWDED'),
 ('gill-size', 'NARROW'),
 ('gill-color', 'WHITE'),
 ('stalk-shape', 'TAPERING'),
 ('stalk-root', 'BULBOUS'),
 ('stalk-surface-above-ring', 'SMOOTH'),
 ('stalk-surface-below-ring', 'SMOOTH'),
 ('stalk-color-above-ring', 'WHITE'),
 ('stalk-color-below-ring', 'WHITE'),
 ('veil-type', 'PARTIAL'),
 ('veil-color', 'WHITE'),
 ('ring-number', 'ONE'),
 ('ring-type', 'PENDANT'),
 ('spore-print-color', 'PURPLE'),
 ('population', 'SEVERAL'),
 ('habitat', 'WOODS')]

In [5]:
# Encoding feature names so they don't get lost along the way having multiple colors with the same values
mushrooms = map(lambda mushroom: map(lambda feat: str(feat)+':'+mushroom[feat], range(len(mushroom))), mushrooms)

In [6]:
# Top 5 mushrooms
for mushroom in mushrooms[:7]:
    print mushroom

['0:EDIBLE', '1:CONVEX', '2:SMOOTH', '3:WHITE', '4:BRUISES', '5:ALMOND', '6:FREE', '7:CROWDED', '8:NARROW', '9:WHITE', '10:TAPERING', '11:BULBOUS', '12:SMOOTH', '13:SMOOTH', '14:WHITE', '15:WHITE', '16:PARTIAL', '17:WHITE', '18:ONE', '19:PENDANT', '20:PURPLE', '21:SEVERAL', '22:WOODS']
['0:EDIBLE', '1:CONVEX', '2:SMOOTH', '3:WHITE', '4:BRUISES', '5:ALMOND', '6:FREE', '7:CROWDED', '8:NARROW', '9:WHITE', '10:TAPERING', '11:BULBOUS', '12:SMOOTH', '13:SMOOTH', '14:WHITE', '15:WHITE', '16:PARTIAL', '17:WHITE', '18:ONE', '19:PENDANT', '20:BROWN', '21:SEVERAL', '22:WOODS']
['0:EDIBLE', '1:CONVEX', '2:SMOOTH', '3:WHITE', '4:BRUISES', '5:ALMOND', '6:FREE', '7:CROWDED', '8:NARROW', '9:PINK', '10:TAPERING', '11:BULBOUS', '12:SMOOTH', '13:SMOOTH', '14:WHITE', '15:WHITE', '16:PARTIAL', '17:WHITE', '18:ONE', '19:PENDANT', '20:PURPLE', '21:SEVERAL', '22:WOODS']
['0:EDIBLE', '1:CONVEX', '2:SMOOTH', '3:WHITE', '4:BRUISES', '5:ALMOND', '6:FREE', '7:CROWDED', '8:NARROW', '9:PINK', '10:TAPERING', '11:BULB

In [7]:
# Performing one hot encoding
oht = OnehotTransactions()
oht_ary = oht.fit(mushrooms).transform(mushrooms)
encoded_shrooms = pd.DataFrame(oht_ary, columns=oht.columns_)
encoded_shrooms.head(7)

0:EDIBLE  0:POISONOUS  10:ENLARGING  10:TAPERING  11:?  11:BULBOUS  \
0         1            0             0            1     0           1   
1         1            0             0            1     0           1   
2         1            0             0            1     0           1   
3         1            0             0            1     0           1   
4         1            0             0            1     0           1   
5         1            0             0            1     0           1   
6         1            0             0            1     0           1   

   11:CLUB  11:EQUAL  11:ROOTED  12:FIBROUS    ...     9:BUFF  9:CHOCOLATE  \
0        0         0          0           0    ...          0            0   
1        0         0          0           0    ...          0            0   
2        0         0          0           0    ...          0            0   
3        0         0          0           0    ...          0            0   
4        0         0          0           0    ...          0            0   
5        0         0          0           0    ...          0            0   
6        0         0          0           0    ...          0            0   

   9:GRAY  9:GREEN  9:ORANGE  9:PINK  9:PURPLE  9:RED  9:WHITE  9:YELLOW  
0       0        0         0       0         0      0        1         0  
1       0        0         0       0         0      0        1         0  
2       0        0         0       1         0      0        0         0  
3       0        0         0       1         0      0        0         0  
4       0        0         0       0         0      0        0         0  
5       0        0         0       0         0      0        0         0  
6       0        0         0       0         0      0        1         0  

[7 rows x 119 columns]

In [8]:
# Listing the available classes
print encoded_shrooms.columns

Index([u'0:EDIBLE', u'0:POISONOUS', u'10:ENLARGING', u'10:TAPERING', u'11:?',
       u'11:BULBOUS', u'11:CLUB', u'11:EQUAL', u'11:ROOTED', u'12:FIBROUS',
       ...
       u'9:BUFF', u'9:CHOCOLATE', u'9:GRAY', u'9:GREEN', u'9:ORANGE',
       u'9:PINK', u'9:PURPLE', u'9:RED', u'9:WHITE', u'9:YELLOW'],
      dtype='object', length=119)


In [9]:
# Executing the apriori algorithm with minimum support of 0.25 and maximal size of frequent itemsets of 9
frequent = apriori(encoded_shrooms, min_support = 0.25, max_len = 9, use_colnames = True)

In [10]:
# Edible and poisonous subsets
frequent['size'] = frequent['itemsets'].apply(lambda x: len(x))
frequent['isEdible'] = frequent['itemsets'].apply(lambda x: '0:EDIBLE' in x)
frequent['isPoison'] = frequent['itemsets'].apply(lambda x: '0:POISONOUS' in x)
frequent.head(10)

support        itemsets  size  isEdible  isPoison
0  0.533270      [0:EDIBLE]     1      True     False
1  0.466730   [0:POISONOUS]     1     False      True
2  0.422053  [10:ENLARGING]     1     False     False
3  0.577947   [10:TAPERING]     1     False     False
4  0.294677          [11:?]     1     False     False
5  0.451521    [11:BULBOUS]     1     False     False
6  0.283270      [12:SILKY]     1     False     False
7  0.631654     [12:SMOOTH]     1     False     False
8  0.273764      [13:SILKY]     1     False     False
9  0.603137     [13:SMOOTH]     1     False     False

### 1. Some itemsets of size 9 that include EDIBLE mushrooms'

In [11]:
# Function to replace field # with actual names
def to_readable(item):
    return map(lambda x: colnames[int(x.split(':')[0])] + ':' + x.split(':')[1], item)

In [12]:
_9edibles = frequent[(frequent['size'] == 9) & (frequent['isEdible'] == True)]

In [13]:
%%capture
_9edibles['itemsets'] = _9edibles['itemsets'].apply(lambda x: to_readable(x))

In [14]:
# Edible itemsets of size 9
_9edibles

support  \
5324  0.266160   
5325  0.266160   
5326  0.277567   
5327  0.266160   
5328  0.266160   
5329  0.277567   
5330  0.266160   
5331  0.266160   
5332  0.266160   
5333  0.266160   
5334  0.266160   
5335  0.266160   
5336  0.269962   
5337  0.269962   
5338  0.269962   
5339  0.269962   
5340  0.292776   
5341  0.266160   
5342  0.266160   
5343  0.266160   
5344  0.266160   
5345  0.266160   
5346  0.269962   
5347  0.266160   
5348  0.266160   
5349  0.266160   
5350  0.266160   
5351  0.266160   
5352  0.269962   
5353  0.266160   
5354  0.288973   
5355  0.288973   
5356  0.288973   
5357  0.288973   
5358  0.288973   
5359  0.292776   
5360  0.288973   
5361  0.288973   
5362  0.266160   
5363  0.266160   
5364  0.266160   
5365  0.266160   
5366  0.266160   
5367  0.269962   
5368  0.266160   
5369  0.266160   
5370  0.288973   

                                                                                                                                                                                                    itemsets  \
5324                         [classes:EDIBLE, stalk-shape:TAPERING, stalk-surface-above-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, odor:NONE, gill-attachment:FREE, gill-size:BROAD]   
5325                         [classes:EDIBLE, stalk-shape:TAPERING, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, odor:NONE, gill-attachment:FREE, gill-size:BROAD]   
5326     [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE]   
5327       [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, bruises?:BRUISES, gill-spacing:CLOSE]   
5328          [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, bruises?:BRUISES, gill-size:BROAD]   
5329   [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, gill-attachment:FREE, gill-spacing:CLOSE]   
5330      [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, gill-attachment:FREE, gill-size:BROAD]   
5331        [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, gill-spacing:CLOSE, gill-size:BROAD]   
5332    [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE]   
5333       [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:BRUISES, gill-attachment:FREE, gill-size:BROAD]   
5334         [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:BRUISES, gill-spacing:CLOSE, gill-size:BROAD]   
5335     [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]   
5336  [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE]   
5337     [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-size:BROAD]   
5338       [classes:ED

In [15]:
%%capture
# Filtering edibles only of all sizes
edibles = frequent[frequent['isEdible'] == True]
edibles['itemsets'] = edibles['itemsets'].apply(lambda x: to_readable(x))

In [16]:
edibles.head(7)

support                                           itemsets  size  \
0   0.533270                                   [classes:EDIBLE]     1   
35  0.338403             [classes:EDIBLE, stalk-shape:TAPERING]     2   
36  0.449144  [classes:EDIBLE, stalk-surface-above-ring:SMOOTH]     2   
37  0.420627  [classes:EDIBLE, stalk-surface-below-ring:SMOOTH]     2   
38  0.360266     [classes:EDIBLE, stalk-color-above-ring:WHITE]     2   
39  0.351711     [classes:EDIBLE, stalk-color-below-ring:WHITE]     2   
40  0.533270                [classes:EDIBLE, veil-type:PARTIAL]     2   

    isEdible  isPoison  
0       True     False  
35      True     False  
36      True     False  
37      True     False  
38      True     False  
39      True     False  
40      True     False

In [17]:
%%capture
# Filtering poisonous only of all sizes
poisonous = frequent[frequent['isPoison'] == True]
poisonous['itemsets'] = poisonous['itemsets'].apply(lambda x: to_readable(x))

In [18]:
poisonous.head(7)

support                                             itemsets  size  \
1   0.466730                                  [classes:POISONOUS]     1   
49  0.266160  [classes:POISONOUS, stalk-surface-above-ring:SILKY]     2   
50  0.256654  [classes:POISONOUS, stalk-surface-below-ring:SILKY]     2   
51  0.466730               [classes:POISONOUS, veil-type:PARTIAL]     2   
52  0.465779                [classes:POISONOUS, veil-color:WHITE]     2   
53  0.452471                 [classes:POISONOUS, ring-number:ONE]     2   
54  0.338403              [classes:POISONOUS, population:SEVERAL]     2   

    isEdible  isPoison  
1      False      True  
49     False      True  
50     False      True  
51     False      True  
52     False      True  
53     False      True  
54     False      True

### 2. Checking if certain mushrooms are edible
### a. A partial white-veiled mushroom with broad, closely spaced gills.

In [19]:
zip(range(len(colnames)), colnames)

[(0, 'classes'),
 (1, 'cap-shape'),
 (2, 'cap-surface'),
 (3, 'cap-color'),
 (4, 'bruises?'),
 (5, 'odor'),
 (6, 'gill-attachment'),
 (7, 'gill-spacing'),
 (8, 'gill-size'),
 (9, 'gill-color'),
 (10, 'stalk-shape'),
 (11, 'stalk-root'),
 (12, 'stalk-surface-above-ring'),
 (13, 'stalk-surface-below-ring'),
 (14, 'stalk-color-above-ring'),
 (15, 'stalk-color-below-ring'),
 (16, 'veil-type'),
 (17, 'veil-color'),
 (18, 'ring-number'),
 (19, 'ring-type'),
 (20, 'spore-print-color'),
 (21, 'population'),
 (22, 'habitat')]

In [20]:
edibles_1 = edibles[['support', 'itemsets']]
poisonous_1 = poisonous[['support', 'itemsets']]

In [21]:
%%capture
# Filtering edibles by mushroom type
edibles_1['filter'] = edibles_1['itemsets'].apply(lambda x: str(colnames[16]+':PARTIAL' in x) + \
                                        str(colnames[17]+':WHITE' in x) + str(colnames[8]+':BROAD' in x) +\
                                        str(colnames[7]+':CLOSE' in x))

# Filtering poisonous by mushroom type
poisonous_1['filter'] = poisonous_1['itemsets'].apply(lambda x: str(colnames[16]+':PARTIAL' in x) + \
                                        str(colnames[17]+':WHITE' in x) + str(colnames[8]+':BROAD' in x) +\
                                        str(colnames[7]+':CLOSE' in x))

In [22]:
edibles_1 = edibles_1[edibles_1['filter'] == 'TrueTrueTrueTrue'][['support', 'itemsets']]
poisonous_1 = poisonous_1[poisonous_1['filter'] == 'TrueTrueTrueTrue'][['support', 'itemsets']]

In [23]:
print "Total number of items = ", len(edibles_1)
edibles_1.head(7)

Total number of items =  57


support  \
2604  0.317490   
3836  0.315589   
3917  0.292776   
4011  0.288973   
4018  0.294677   
4021  0.315589   
4024  0.317490   

                                                                                                                         itemsets  
2604                                   [classes:EDIBLE, veil-type:PARTIAL, veil-color:WHITE, gill-spacing:CLOSE, gill-size:BROAD]  
3836  [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, gill-spacing:CLOSE, gill-size:BROAD]  
3917  [classes:EDIBLE, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, gill-spacing:CLOSE, gill-size:BROAD]  
4011                  [classes:EDIBLE, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, gill-spacing:CLOSE, gill-size:BROAD]  
4018                [classes:EDIBLE, veil-type:PARTIAL, veil-color:WHITE, ring-type:PENDANT, gill-spacing:CLOSE, gill-size:BROAD]  
4021                 [classes:EDIBLE, veil-type:PARTIAL, veil-color:WHITE, bruises?:BRUISES, gill-spacing:CLOSE, gill-size:BROAD]  
4024             [classes:EDIBLE, veil-type:PARTIAL, veil-color:WHITE, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]

In [24]:
edibles_1.tail(3)

support  \
5366  0.266160   
5367  0.269962   
5370  0.288973   

                                                                                                                                                                                    itemsets  
5366    [classes:EDIBLE, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]  
5367  [classes:EDIBLE, stalk-surface-below-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]  
5370                  [classes:EDIBLE, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]

In [25]:
print "Total number of items = ", len(poisonous_1)
poisonous_1.head(7)

Total number of items =  0


Empty DataFrame
Columns: [support, itemsets]
Index: []

In [26]:
poisonous_1.tail(3)

Empty DataFrame
Columns: [support, itemsets]
Index: []

### Running the apriori algorithm for frequent itemsets that have support of 0.25 and have no more than 9 items I have the following results for partial white-veiled mushrooms with broad, closely spaced gills:
### 1. The support for the edible mushrooms of this type is 0.317490. There are 57 different itemsets having this type with the minimal support of 0.288973.
### 2. There are no poisonous mushrooms of this type.
### Therefore, I conclude that the mushrooms are edible.

### b. A partial white-veiled mushroom with closely spaced gills, a single ring and a silky surface below the ring.

In [27]:
zip(range(len(colnames)), colnames)

[(0, 'classes'),
 (1, 'cap-shape'),
 (2, 'cap-surface'),
 (3, 'cap-color'),
 (4, 'bruises?'),
 (5, 'odor'),
 (6, 'gill-attachment'),
 (7, 'gill-spacing'),
 (8, 'gill-size'),
 (9, 'gill-color'),
 (10, 'stalk-shape'),
 (11, 'stalk-root'),
 (12, 'stalk-surface-above-ring'),
 (13, 'stalk-surface-below-ring'),
 (14, 'stalk-color-above-ring'),
 (15, 'stalk-color-below-ring'),
 (16, 'veil-type'),
 (17, 'veil-color'),
 (18, 'ring-number'),
 (19, 'ring-type'),
 (20, 'spore-print-color'),
 (21, 'population'),
 (22, 'habitat')]

In [28]:
edibles_2 = edibles[['support', 'itemsets']]
poisonous_2 = poisonous[['support', 'itemsets']]

In [29]:
%%capture
# Filtering edibles by mushroom type
edibles_2['filter'] = edibles_2['itemsets'].apply(lambda x: str(colnames[16]+':PARTIAL' in x) + \
                                        str(colnames[17]+':WHITE' in x) + str(colnames[7]+':CLOSE' in x) + \
                                        str(colnames[18]+':ONE' in x) + str(colnames[13]+':SILKY' in x)) 

# Filtering poisonous by mushroom type
poisonous_2['filter'] = poisonous_2['itemsets'].apply(lambda x: str(colnames[16]+':PARTIAL' in x) + \
                                        str(colnames[17]+':WHITE' in x) + str(colnames[7]+':CLOSE' in x) + \
                                        str(colnames[18]+':ONE' in x) + str(colnames[13]+':SILKY' in x)) 

In [30]:
edibles_2 = edibles_2[edibles_2['filter'] == 'TrueTrueTrueTrueTrue'][['support', 'itemsets']]
poisonous_2 = poisonous_2[poisonous_2['filter'] == 'TrueTrueTrueTrueTrue'][['support', 'itemsets']]

In [31]:
print "Total number of items = ", len(edibles_2)
edibles_2.head(7)

Total number of items =  0


Empty DataFrame
Columns: [support, itemsets]
Index: []

In [32]:
edibles_2.tail(3)

Empty DataFrame
Columns: [support, itemsets]
Index: []

In [33]:
print "Total number of items = ", len(poisonous_2)
poisonous_2.head(7)

Total number of items =  4


support  \
4080  0.256654   
4926  0.256654   
4927  0.256654   
5271  0.256654   

                                                                                                                                                              itemsets  
4080                                     [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, gill-spacing:CLOSE]  
4926                        [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:NO, gill-spacing:CLOSE]  
4927               [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, gill-attachment:FREE, gill-spacing:CLOSE]  
5271  [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:NO, gill-attachment:FREE, gill-spacing:CLOSE]

In [34]:
poisonous_2.tail(3)

support  \
4926  0.256654   
4927  0.256654   
5271  0.256654   

                                                                                                                                                              itemsets  
4926                        [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:NO, gill-spacing:CLOSE]  
4927               [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, gill-attachment:FREE, gill-spacing:CLOSE]  
5271  [classes:POISONOUS, stalk-surface-below-ring:SILKY, veil-type:PARTIAL, veil-color:WHITE, ring-number:ONE, bruises?:NO, gill-attachment:FREE, gill-spacing:CLOSE]

### Running the apriori algorithm for frequent itemsets that have support of 0.25 and have no more than 9 items I have the following results for partial white-veiled mushroom with closely spaced gills, a single ring and a silky surface below the ring:
### 1. There are no frequent itemsets with support of 0.25+ for edible mushrooms of that type. 
### 2. Yet there is 0.256654 support for Poisonous mushrooms of this type and there are 4 frequent itemset with support > 0.25. Therefore, I conclude that this mushroom type is poisonous, i.e. inedible.

### c. A mushroom with smooth surface above the ring.

In [35]:
zip(range(len(colnames)), colnames)

[(0, 'classes'),
 (1, 'cap-shape'),
 (2, 'cap-surface'),
 (3, 'cap-color'),
 (4, 'bruises?'),
 (5, 'odor'),
 (6, 'gill-attachment'),
 (7, 'gill-spacing'),
 (8, 'gill-size'),
 (9, 'gill-color'),
 (10, 'stalk-shape'),
 (11, 'stalk-root'),
 (12, 'stalk-surface-above-ring'),
 (13, 'stalk-surface-below-ring'),
 (14, 'stalk-color-above-ring'),
 (15, 'stalk-color-below-ring'),
 (16, 'veil-type'),
 (17, 'veil-color'),
 (18, 'ring-number'),
 (19, 'ring-type'),
 (20, 'spore-print-color'),
 (21, 'population'),
 (22, 'habitat')]

In [36]:
edibles_3 = edibles[['support', 'itemsets']]
poisonous_3 = poisonous[['support', 'itemsets']]

In [37]:
%%capture
# Filtering edibles by mushroom type
edibles_3['filter'] = edibles_3['itemsets'].apply(lambda x: str(colnames[12]+':SMOOTH' in x))

# Filtering poisonous by mushroom type
poisonous_3['filter'] = poisonous_3['itemsets'].apply(lambda x: str(colnames[12]+':SMOOTH' in x))

In [38]:
edibles_3 = edibles_3[edibles_3['filter'] == 'True'][['support', 'itemsets']]
poisonous_3 = poisonous_3[poisonous_3['filter'] == 'True'][['support', 'itemsets']]

In [39]:
print "Total number of items = ", len(edibles_3)
edibles_3.head(7)

Total number of items =  646


support  \
36   0.449144   
278  0.277567   
286  0.379515   
287  0.278042   
288  0.273764   
289  0.449144   
290  0.426331   

                                                                               itemsets  
36                                    [classes:EDIBLE, stalk-surface-above-ring:SMOOTH]  
278             [classes:EDIBLE, stalk-shape:TAPERING, stalk-surface-above-ring:SMOOTH]  
286  [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-surface-below-ring:SMOOTH]  
287     [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-color-above-ring:WHITE]  
288     [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, stalk-color-below-ring:WHITE]  
289                [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, veil-type:PARTIAL]  
290                 [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, veil-color:WHITE]

In [40]:
edibles_3.tail(3)

support  \
5359  0.292776   
5360  0.288973   
5361  0.288973   

                                                                                                                                                                                    itemsets  
5359  [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, veil-type:PARTIAL, veil-color:WHITE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]  
5360   [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, veil-type:PARTIAL, ring-number:ONE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]  
5361    [classes:EDIBLE, stalk-surface-above-ring:SMOOTH, veil-color:WHITE, ring-number:ONE, ring-type:PENDANT, bruises?:BRUISES, gill-attachment:FREE, gill-spacing:CLOSE, gill-size:BROAD]

In [41]:
print "Total number of items = ", len(poisonous_3)
poisonous_3.head(7)

Total number of items =  0


Empty DataFrame
Columns: [support, itemsets]
Index: []

In [42]:
poisonous_3.tail(3)

Empty DataFrame
Columns: [support, itemsets]
Index: []

### Running the apriori algorithm for frequent itemsets that have support of 0.25 and have no more than 9 items I have the following results for mushrooms with smooth surface above the ring.:
### 1. The support for the edible mushrooms of this type is 0.449144. There are 646 different itemsets having this type with the minimal support of 0.288973.
### 2. There are no poisonous mushrooms of this type.
### Therefore, I conclude that the mushrooms are edible.